compiler/sem/sem.nim

```diff
@@ -351,6 +351,120 @@ proc newSymS(kind: TSymKind, n: PNode, c: PContext): PSym =
   when defined(nimsuggest):
     suggestDecl(c, n, result)
 
+func defNameErrorNodeAllowsSymUpdate*(n: PNode): bool {.inline.} =
+  ## true if `n` is an `nkError` of kind `adSemSymNameSym` and the error is
+  ## minor enough to allow the recovery sym to be updated. Used to see if
+  ## progress is allowed inspite of errors.
+  n.kind == nkError and n.diag.kind == adSemDefNameSym and
+    n.diag.defNameSymData.kind == adSemDefNameSymExpectedKindMismatch
+
+func getDefNameSymOrRecover*(n: PNode): PSym {.inline.} =
+  ## extracts the symbol from `n`, which must be an `nkSym` or `nkError` with
+  ## diagnostic kind of `adSemDefNameSym`. If `n` is an error, a
+  ## recovery symbol is extracted, allowing progress to be made.
+  case n.kind
+  of nkSym: n.sym
+  of nkError:
+    case n.diag.kind
+    of adSemDefNameSym: n.diag.defNameSym
+    else: unreachable("all error cases must be covered")
+  else:
+    unreachable("no other cases supported")
+
+proc newSymGNode*(kind: TSymKind, n: PNode, c: PContext): PNode =
+  ## like newSymS, but considers gensym'ed symbols, analyses `n` producing a
+  ## canonical symbol node (`nkSym`), or if unsuccessful an `nkError` with an
+  ## `adSemDefNameSym` diagnostic.
+  ## 
+  ## Symbol canonicalization is is as follows:
+  ## 1. `nkSym` -> match `kind` (or `skTemp`) and updates the sym owner to the
+  ##               current context (gensym specific)
+  ## 2. `nkIdentKinds` -> instantiate a symbol considering quoted identifiers
+  ## 3. `nkError` -> simply returns the same error
+  ## 4. other kinds -> new error
+  # TODO: come up with a better name for this proc
+  let
+    info = n.info
+    currOwner = getCurrOwner(c)
+
+  template makeError(config: ConfigRef, n: PNode, sym: PSym,
+                     dataKindSuffix: untyped): PNode =
+    let
+      dataKind = `adSemDefNameSym dataKindSuffix`
+      (wrongNode, defData) =
+        case dataKind
+        of adSemDefNameSymExpectedKindMismatch:
+          (n, AdSemDefNameSym(kind: dataKind, expectedKind: kind))
+        of adSemDefNameSymIdentGenFailed:
+          (n.diag.wrongNode,
+          AdSemDefNameSym(kind: dataKind, identGenErr: n))
+        of adSemDefNameSymExistingError,
+            adSemDefNameSymIllformedAst:
+          (n, AdSemDefNameSym(kind: dataKind))
+
+    config.newError(wrongNode, PAstDiag(kind: adSemDefNameSym,
+                                        defNameSym: sym,
+                                        defNameSymData: defData))
+
+  case n.kind
+  of nkSym:
+    result =
+      # xxx: why don't we guard on `sfGenSym`?
```

Comment:
Yup, which works in the other direction too, where they could transplant an otherwise valid definitional `nkSym` into another definition spot and it's _really_ hard to tell that that happened.